File: leetcode/c++/86_partition_list.cpp
// https://leetcode.cn/problems/partition-list/description

#include<bits/stdc++.h>
using namespace std;


struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode* list1 = new ListNode();
        ListNode* list2 = new ListNode();
        ListNode* end1 = list1;
        ListNode* end2 = list2;
        while (head != nullptr) {
            if (head->val < x) {              
                end1->next = head;
                end1 = end1->next;              
            } else {
                end2->next = head;
                end2 = end2->next;
            }
            head = head->next;
        }
        end2->next = nullptr;
        end1->next = list2->next;
        ListNode* res = list1->next;
        delete list1;
        delete list2;
        return res;
    }
};

// 时间复杂度为O(n)，空间复杂度为O(1)

File: leetcode/golang/86_partition_list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func partition(head *ListNode, x int) *ListNode {
    list1 := ListNode {}
    list2 := ListNode {}
    end1 := &list1
    end2 := &list2
    for head != nil {
        if head.Val < x {
            end1.Next = head
            end1 = end1.Next
        } else {
            end2.Next = head
            end2 = end2.Next
        }
        head = head.Next
    }
    end1.Next = list2.Next
    end2.Next = nil
    return list1.Next
}